libvast/src/plugin.cpp

```diff
@@ -40,18 +48,188 @@ std::string to_string(plugin_version x) {
+  // Step 1: Get the necessary options.
+  auto paths_or_names
+    = caf::get_or(cfg, "vast.plugins", std::vector<std::string>{});
+  if (paths_or_names.empty())
+    return loaded_plugin_paths;
+  const auto plugin_dirs = get_plugin_dirs(cfg);
+  // Step 2: Try to resolve the reserved identifier 'all'.
+  if (const auto all
+      = std::remove(paths_or_names.begin(), paths_or_names.end(), "all");
+      all != paths_or_names.end()) {
+    paths_or_names.erase(all, paths_or_names.end());
+    for (const auto& dir : plugin_dirs) {
+      auto ec = std::error_code{};
+      for (const auto& file : std::filesystem::directory_iterator{dir, ec}) {
+        if (ec || !file.is_regular_file())
+          break;
+        if (detail::starts_with(file.path().filename().string(), //
+                                "libvast-plugin-"))
+          paths_or_names.push_back(file.path());
+      }
+    }
+  }
+  // Step 3: Try to resolve the reserved identifier 'bundled'.
+  if (const auto bundled
+      = std::remove(paths_or_names.begin(), paths_or_names.end(), "bundled");
+      bundled != paths_or_names.end()) {
+    paths_or_names.erase(bundled, paths_or_names.end());
+    std::move(bundled_plugins.begin(), bundled_plugins.end(),
+              std::back_inserter(paths_or_names));
+  }
+  // Step 4: Disable static plugins that were not enabled.
+  auto is_disabled_static_plugin = [&](auto& plugin) -> bool {
+    switch (plugin.type()) {
+      case plugin_ptr::type::dynamic:
+        die("dynamic plugins must not be loaded at this point");
+      case plugin_ptr::type::static_: {
+        auto has_same_name = [&](const auto& name) {
+          return plugin->name() == name;
+        };
+        return std::none_of(paths_or_names.begin(), paths_or_names.end(),
+                            has_same_name);
+      }
+      case plugin_ptr::type::native:
+        return false;
+    }
+    die("unreachable");
+  };
+  get_mutable().erase(std::remove_if(get_mutable().begin(), get_mutable().end(),
+                                     is_disabled_static_plugin),
+                      get_mutable().end());
+  // Step 5: Try to resolve plugin names to plugin paths.
+  for (auto& path_or_name : paths_or_names) {
+    // Ignore paths.
+    if (auto maybe_path = std::filesystem::path{path_or_name};
+        maybe_path.is_absolute())
+      continue;
+    // At this point, we only have names—that we need to resolve to
+    // `{dir}/libvast-plugin-{name}.{ext}`. We take the first file that
+    // exists.
+    if (auto path = resolve_plugin_name(plugin_dirs, path_or_name))
+      path_or_name = path->string();
+    else
+      return std::move(path.error());
+  }
+  // Step 6: Deduplicate plugin paths.
+  auto paths = detail::stable_set<std::string>{};
+  paths.insert(std::make_move_iterator(paths_or_names.begin()),
+               std::make_move_iterator(paths_or_names.end()));
+  // Step 7: Load plugins.
```

Comment:
I think here would be a natural place to split up the function (with step 4 pulled out) since the parts before and after are unrelated; one function to resolve the list of plugins to be loaded and one that does the loading + ordering.

Reply:
I'm not sure how to extract step 4 without duplicating some of its logic in later steps. Let's talk tomorrow about this. Other than that I've incorporated all feedback.